composition/src/index.js

```diff
@@ -0,0 +1,260 @@
+		c.__compositions.w.forEach(up => handleEffect(up, c));
+	}
+};
+
+let oldAfterDiff = options.diffed;
+options.diffed = vnode => {
+	if (oldAfterDiff) oldAfterDiff(vnode);
+
+	/** @type {import('./internal').Component} */
+	const c = vnode._component;
+	if (c && c.__compositions)
+		// handle all `effect`s
+		c.__compositions.e.forEach(up => handleEffect(up, c));
+};
+
+let oldBeforeUnmount = options.unmount;
+options.unmount = vnode => {
+	if (oldBeforeUnmount) oldBeforeUnmount(vnode);
+
+	/** @type {import('./internal').Component} */
+	const c = vnode._component;
+	if (c && c.__compositions) {
+		// cleanup `effect`s onCleanup
+		c.__compositions.e.forEach(cleanupEffect);
+		// call all onUnmounted lifecycle callbacks
+		c.__compositions.u.forEach(f => f());
+	}
+};
+
+const $Reactive = Symbol('reactive');
+
+export function createComponent(comp) {
+	comp.__compositions = true;
+	return comp;
+}
+
+export function watch(src, cb, dv) {
+	const vr = { value: dv };
+	Object.defineProperty(vr, $Reactive, {
+		get() {
+			return this.value;
+		}
+	});
+	const up = { src, cb, vr };
+	handleEffect(up, currentComponent);
+	currentComponent.__compositions.w.push(up);
+	return vr;
+}
+
+export function effect(src, cb) {
+	currentComponent.__compositions.e.push({ src, cb });
+}
+
+export function onMounted(cb) {
+	currentComponent._renderCallbacks.push(cb);
+}
+
+export function onUnmounted(cb) {
+	currentComponent.__compositions.u.push(cb);
+}
+
+export function provide(name, _value) {
+	const c = currentComponent;
+	if (!c.__compositions.c) {
+		c.__compositions.c = {};
+		c.getChildContext = () => c.__compositions.c;
+	}
+	c.__compositions.c[`__sC_${name}`] = { _component: c, _value };
+}
+
+export function inject(name, defaultValue) {
+	const c = currentComponent;
+
+	const ctx = c.context[`__sC_${name}`];
+	if (!ctx) return defaultValue;
+
+	const src = ctx._value;
+
+	if (isReactive(src))
+		ctx._component.__compositions.w.push({
+			src,
+			cb: () => c.setState(EMPTY_STATE)
+		});
+
+	return src;
+}
+
+export function reactive(value) {
+	let x = value;
+	const c = currentComponent;
+
+	const reactiveProperty = {
+		get() {
+			return x;
+		},
+		set(v) {
+			x = v;
+			c.setState(EMPTY_STATE);
+		}
+	};
+	return Object.defineProperties(
+		{},
+		Object.keys(value).reduce(
+			(acc, key) =>
+				Object.assign(acc, {
+					[key]: {
+						enumerable: true,
+						get() {
+							return x[key];
+						},
+						set(v) {
+							if (v !== x[key]) {
+								x = Object.assign({}, x);
+								x[key] = v;
+								c.setState(EMPTY_STATE);
+							}
+						}
+					}
+				}),
+			{
+				[$Reactive]: reactiveProperty,
+				$value: reactiveProperty
+			}
+		)
+	);
+}
+
+export function value(v) {
+	const c = currentComponent;
+	function get() {
+		return v;
+	}
+	return Object.defineProperties(
+		{},
+		{
+			[$Reactive]: { get },
+			value: {
+				get,
+				set(newValue) {
+					if (v !== newValue) {
+						v = newValue;
+						c.setState(EMPTY_STATE);
```

Comment:
You only use this once, I'd inline it

Reply:
At the point this function is called here, `currentComponent` may have already change, so i need to keep a reference to the component this belongs to.
It's kinda like what we do with hooks, but hooks are called every render, so `currentComponent` is the component the hook refers too, while compositions are only called once at setup phase